datafusion/expr-common/src/signature.rs

```diff
@@ -243,6 +243,27 @@ impl TypeSignature {
             _ => false,
         }
     }
+
+    /// get all possible types for the given `TypeSignature`
+    pub fn get_possible_types(&self) -> Vec<Vec<DataType>> {
+        match self {
+            TypeSignature::Exact(types) => vec![types.clone()],
+            TypeSignature::OneOf(types) => types
+                .iter()
+                .flat_map(|type_sig| type_sig.get_possible_types())
+                .collect(),
+            // TODO: Implement for other types
+            TypeSignature::Uniform(_, _)
+            | TypeSignature::Coercible(_)
```

Comment:
There is a follow-up issue for inferring the types from other TypeSignatures.

Reply:
I think we should file a ticket to track adding the feature (I bet some other contributors would enjoy helping out now that you have setup the framework)

Reply:
Sounds great. I filed https://github.com/apache/datafusion/issues/13271 for it.